src/server/system_services/bucket_server.js

```diff
@@ -1900,6 +1905,112 @@ function validate_non_nsfs_bucket_creation(req) {
+
+    return bucket_names_replication;
+}
+
+async function delete_bucket_replication(req) {
+    dbg.log0('delete_bucket_replication:', req.rpc_params);
+    const bucket = find_bucket(req);
+    const replication_id = bucket.replication_policy_id;
+    if (!replication_id) return;
+    await system_store.make_changes({
+        update: {
+            buckets: [{
+                _id: bucket._id,
+                $unset: { replication_policy_id: 1 }
+            }]
+        }
+    });
+
+    // delete replication from replication collection
+    await replication_store.instance().delete_replication_by_id(replication_id);
+}
+
+function validate_replication(req, bucket_name, min_required) {
+    const replication_rules = req.rpc_params.replication_policy;
+    // num of rules in configuration must be in defined limits
+    if (replication_rules.length > config.BUCKET_REPLICATION_MAX_RULES ||
+        (min_required && replication_rules.length < 1)) throw new RpcError('INVALID_REPLICATION_POLICY');
+
+    let rule_ids = [];
+    let pref_by_dst_bucket = {};
+
+    _.map(replication_rules, rule => {
+        const { destination_bucket, filter, rule_id } = rule;
+        const dst_bucket = req.system.buckets_by_name && req.system.buckets_by_name[destination_bucket.unwrap()];
+        // rule's destination bucket must exist and not equal to the replicated bucket
+        if (bucket_name === destination_bucket.unwrap() || !dst_bucket) {
+            throw new RpcError('INVALID_REPLICATION_POLICY');
+        }
+
+        rule_ids.push(rule_id);
+        pref_by_dst_bucket[destination_bucket] = pref_by_dst_bucket[destination_bucket] || [];
+        pref_by_dst_bucket[destination_bucket].push((filter && filter.prefix) || '');
+        return { ...rule, destination_bucket: dst_bucket._id };
+    });
+    // all rule_ids must be different
+    if (new Set(rule_ids).size < replication_rules.length) throw new RpcError('INVALID_REPLICATION_POLICY');
+
+    // num of different destination buckets in configuration must be in defined limits
+    if (Object.keys(pref_by_dst_bucket).length > config.BUCKET_REPLICATION_MAX_DST_BUCKETS) {
+        throw new RpcError('INVALID_REPLICATION_POLICY');
+    }
+    // all prefixes of same destination bucket must not be a prefix of another prefix.
+    for (const dst in pref_by_dst_bucket) {
+        if (pref_by_dst_bucket[dst].length > 1) {
+            const ordered_prefixes = _.orderBy(pref_by_dst_bucket[dst]);
+            for (let i = 0; i < ordered_prefixes.length + 1; i++) {
+                if (ordered_prefixes[i + 1].startsWith(ordered_prefixes[i])) {
+                    throw new RpcError('INVALID_REPLICATION_POLICY');
+                }
+            }
+        }
+    }
+}
+
+function normalize_replication(req) {
+    const replication_rules = req.rpc_params.replication_policy;
+
+    const validated_replication = _.map(replication_rules, rule => {
```

Comment:
since `replication_rules` is an array I think it's more readable to use `replication_rules.map`
```suggestion
    const validated_replication = replication_rules.map(rule => {
```